src/libraries/System.Data.Common/src/System/Data/Common/DbDataSource.cs

```diff
@@ -0,0 +1,497 @@
+{
+    public abstract class DbDataSource : IDisposable, IAsyncDisposable
+    {
+        public abstract string ConnectionString { get; }
+
+        protected abstract DbConnection CreateDbConnection();
+
+        protected virtual DbConnection OpenDbConnection()
+        {
+            var connection = CreateDbConnection();
+            connection.Open();
+
+            return connection;
+        }
+
+        protected virtual async ValueTask<DbConnection> OpenDbConnectionAsync(CancellationToken cancellationToken = default)
+        {
+            var connection = CreateDbConnection();
+            await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+
+            return connection;
+        }
+
+        protected virtual DbCommand CreateDbCommand(string? commandText = null)
+        {
+            var command = CreateDbConnection().CreateCommand();
+            command.CommandText = commandText;
+
+            return new DbCommandWrapper(command);
+        }
+
+        protected virtual DbBatch CreateDbBatch()
+            => new DbBatchWrapper(CreateDbConnection().CreateBatch());
+
+        public DbConnection CreateConnection() => CreateDbConnection();
+        public DbConnection OpenConnection() => OpenDbConnection();
+        public ValueTask<DbConnection> OpenConnectionAsync() => OpenDbConnectionAsync();
+        public DbCommand CreateCommand(string? commandText = null) => CreateDbCommand();
+        public DbBatch CreateBatch() => CreateDbBatch();
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            await DisposeAsyncCore().ConfigureAwait(false);
+
+            Dispose(disposing: false);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        protected virtual ValueTask DisposeAsyncCore()
+            => default;
+
+        private sealed class DbCommandWrapper : DbCommand
+        {
+            private readonly DbCommand _wrappedCommand;
+            private readonly DbConnection _connection;
+
+            internal DbCommandWrapper(DbCommand wrappedCommand)
+            {
+                Debug.Assert(wrappedCommand.Connection is not null);
+
+                _wrappedCommand = wrappedCommand;
+                _connection = wrappedCommand.Connection;
+            }
+
+            public override int ExecuteNonQuery()
+            {
+                _connection.Open();
+
+                try
+                {
+                    return _wrappedCommand.ExecuteNonQuery();
+                }
+                finally
+                {
+                    try
+                    {
+                        _connection.Close();
+                    }
+                    catch
+                    {
+                        // Swallow to allow the original exception to bubble up
```

Comment:
There might not be an original exception.

Reply:
That's true - I had an offline discussion with @NinoFloris about whether Close exceptions should generally bubble up.

On the one hand, Close failures aren't really relevant to the execution itself, which completed successfully - which is why the current code doesn't (ever) bubble them up (the idea behind this API is that the user doesn't care about connection-related details any more). It's a bit comparable to what happens if an internal logging failure occurs in some library; usually you don't want that to be bubbled up to the user, since logging is secondary and failure there doesn't affect the main operation.

On the other hand, Open exceptions do bubble up (since we can't execute), so it's maybe somewhat inconsistent.

I'm still leaning to not bubbling it up (comment could be updated to reflect the reasoning), but am interested in opinions to the contrary.

Reply:
Should it log the swallowed exceptions to DataCommonEventSource? There is ExceptionBuilder.TraceExceptionWithoutRethrow for that purpose, but it seems to be used in DataTable-related classes only.

Reply:
Good suggestion... I suspect it's only currently used in DataTable-related classes is that the main System.Data classes are abstractions only, and therefore don't need to swallow/report exceptions. I'll make that change - it would allow visibility to anyone who's really interested in these exceptions, and I don't see a downside to it.

Reply:
The commit message is "Trace swallowed exceptions and fix comments" but it doesn't add any tracing.

Reply:
Pushed it